Implement account service (mediatr, cqrs)

Yarashine opened the pull request:
Implement the account service without authorisation, as this functionality should be implemented after the interaction is configured

Yarashine committed:
Implement account service (mediatr, cqrs)

DarkKezar commented on src/backend/AccountService/AccountService.API/Controllers/AccountController.cs:
```diff
@@ -0,0 +1,83 @@
+    [HttpPost]
+    public async Task<IActionResult> CreateAccount(CancellationToken cancellationToken = default)
+    {
+        //var userId = User.FindFirst(JwtRegisteredClaimNames.Sub)?.Value ?? throw new UnauthorizedAccessException("User is unauthorized");
```
Unused comments shouldn't be in result commit.

DarkKezar commented on src/backend/AccountService/AccountService.API/Controllers/AccountController.cs:
```diff
@@ -0,0 +1,83 @@
+    {
+        //var userId = User.FindFirst(JwtRegisteredClaimNames.Sub)?.Value ?? throw new UnauthorizedAccessException("User is unauthorized");
+        //var convertedId = int.Parse(userId);
+        var convertedId = 0;
```
This is magic number. If you wanna send it to the CreateAccountCommandDto - you can use default value.

DarkKezar commented on src/backend/AccountService/AccountService.API/Program.cs:
```diff
@@ -7,20 +18,46 @@ public static void Main(string[] args)
             builder.Services.AddControllers();
             // Learn more about configuring OpenAPI at https://aka.ms/aspnet/openapi
-            builder.Services.AddOpenApi();
+            builder.Services.AddDbContext<AccountDbContext>(options =>
```
Use extension methods to right declaration of services dependencies

DarkKezar commented on src/backend/AccountService/AccountService.API/Program.cs:
```diff
@@ -7,20 +18,46 @@ public static void Main(string[] args)
+                });
             }
 
+            using (var scope = app.Services.CreateScope())
```
Data base DI should be declared early than DI of services that's depends on database

Yarashine replied:
Can you explain this comment? I think I didn't declare DI in this line

DarkKezar commented on src/backend/AccountService/AccountService.BLL/UseCases/Account/Commands/Delete/DeleteAccountCommandHandler.cs:
```diff
@@ -0,0 +1,14 @@
+
+public class DeleteAccountCommandHandler(IAccountRepository accountRepository) : IRequestHandler<DeleteAccountCommand>
+{
+    public async Task Handle(DeleteAccountCommand request, CancellationToken cancellationToken = default)
```
I prefer, when Handlers return result (status, or done work)

Yarashine replied:
I used this approach earlier, but I have never use this information. I will have false return if there is exception, but exceptions are already handled

DarkKezar commented on src/backend/AccountService/AccountService.BLL/UseCases/Account/Commands/Transfer/TransferAccountCommandHandler.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public async Task Handle(TransferAccountCommand request, CancellationToken cancellationToken = default)
+    {
+        var accountForCheckSender = await accountRepository.GetByIdAsync(request.SenderAccountId, cancellationToken) ?? throw new NotFoundException("Account");
```
?? operand in such situation make code less readable. Rewrite as several lines.

DarkKezar commented on src/backend/AccountService/AccountService.BLL/UseCases/Account/Commands/Update/UpdateAccountCommandHandler.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace AccountService.BLL.UseCases.Account.Commands.Update;
+
+public class UpdateAccountCommandHandler(
```
When you use default constructor, you cannot control access modifiers of fields. Rewrite to full declaration with full constructor.

DarkKezar commented on src/backend/AccountService/AccountService.DAL/Contracts/Repositories/IAccountRepository.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace AccountService.DAL.Contracts.Repositories;
+
+public interface IAccountRepository
```
Pattern repository contains common abstraction IRepository<T>, that contains default methods (CRUD)

DarkKezar commented on src/backend/AccountService/AccountService.DAL/Data/AccountDbContext.cs:
```diff
@@ -0,0 +1,21 @@
+
+    protected override void OnModelCreating(ModelBuilder modelBuilder)
+    {
+        modelBuilder.Entity<Account>()
```
Not critical in this case. But model creation properties are in specific extensions method commonly.

DarkKezar commented on src/backend/AccountService/AccountService.DAL/Entities/Account.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace AccountService.DAL.Entities;
+
+public class Account
```
Your models can be connected with abstraction "Entity" that will contain common fields.

Yarashine replied:
Why can I use this approach?

DarkKezar commented on src/backend/AccountService/AccountService.DAL/Repositories/AccountRepository.cs:
```diff
@@ -0,0 +1,63 @@
+
+namespace AccountService.DAL.Repositories;
+
+public class AccountRepository(AccountDbContext dbContext) : IAccountRepository
```
I have already wrote about IRepository<T>. Also pattern realisation default contain BaseRepository<T>, this is abstract class with base implementation of IRepository<T>.
Implementation of specific repository should be in next form: **_AccountRepository: BaseRepository<T>, IAccountRepository_**

Where **_BaseRepository: IRepository<T>_** and **_IAccountRepository: IRepository<Account>_**

DarkKezar commented on src/backend/AccountService/AccountService.DAL/Repositories/AccountRepository.cs:
```diff
@@ -0,0 +1,63 @@
+    public DbSet<Account> accounts = dbContext.Accounts;
+    public DbSet<Transfer> transfers = dbContext.Transfers;
+
+    public async Task CreateTransferAsync(Transfer transfer, CancellationToken cancellationToken = default)
```
A lot of business logic in repository method.

DarkKezar commented on src/backend/AccountService/AccountService.DAL/Repositories/AccountRepository.cs:
```diff
@@ -0,0 +1,63 @@
+            .ToListAsync(cancellationToken);
+    }
+
+    public async Task<IEnumerable<Transfer>> GetAllTransfersByIdAsync(Guid id, CancellationToken cancellationToken = default)
```
Probably should be in TransferRepository. Also GetAll methods should contain filtering and pagination options to avoid downloading all data from database (read about Expression<>)

Yarashine committed:
Trying fix swagger

Yarashine committed:
Fixing pull account request first iteration

FaiTh725 commented on src/backend/AccountService/AccountService.DAL/Repositories/TransferRepository.cs:
```diff
@@ -0,0 +1,39 @@
+    public DbSet<Transfer> transfers = dbContext.Transfers;
+    public async Task CreateTransferAsync(Transfer transfer, CancellationToken cancellationToken = default)
+    {
+        await transfers.AddAsync(transfer, cancellationToken);
+
+        var sender = await accounts
+            .FirstOrDefaultAsync(a => a.Id == transfer.SenderAccountId, cancellationToken);
+        sender.Balance -= transfer.Amount;
+        sender.UpdatedAt = DateTime.UtcNow;
+
+        var reciever = await accounts
+            .FirstOrDefaultAsync(a => a.Id == transfer.ReceiverAccountId, cancellationToken);
+        reciever.Balance += transfer.Amount;
+        reciever.UpdatedAt = DateTime.UtcNow;
```
Repositories shouldnt contain business logic

anticlown322 commented on src/backend/AccountService/AccountService.API/Controllers/AccountController.cs:
```diff
@@ -0,0 +1,54 @@
+        var command = new CreateAccountCommand { UserId = userId };
+        await mediator.Send(command, cancellationToken);
+        return Ok();
+    }
```
Remember to add a blank line before a `return` statement. It makes code more readable and meets team code style.
### Proposed solution 
```suggestion
    public async Task<IActionResult> CreateAccount(int userId, CancellationToken cancellationToken = default)
    {
        var command = new CreateAccountCommand { UserId = userId };
        await mediator.Send(command, cancellationToken);
        
        return Ok();
    }
```

anticlown322 commented on src/backend/AccountService/AccountService.API/appsettings.json:
```diff
@@ -5,5 +5,11 @@
+  "Pagination": {
+    "DefaultPageSize": 3
+  },
+  "ConnectionStrings": {
+    "MSSQL": "Server=mssql,1433;Database=MSSQLDB;User Id=sa;Password=MyStrong@Passw0rd;TrustServerCertificate=True;"
+  },
```
Consider to move sensitive data from `appsettings.json` file to a more secure location, such as docker secrets, external config files or env. variables.
### Proposed solution
This solution explains a way with environment variables and docker-compose:
1. Create a file with environment variables in the project root. 
Example path: `\src\backend\AccountSerivce\.env`
Remember to add this file to `.gitignore`. Consider adding a `.env.example` template to the repo with placeholder values
2. Move sensetive data to the creaeted file:
```
ACCOUNT_SERVICE_CONNECTION=Server=mssql,1433;Database=MSSQLDB;User Id=sa;Password=MyStrong@Passw0rd;TrustServerCertificate=True;
```
3. Update `docker-compose.yaml`:
```
 accountservice.api:
    image: ${DOCKER_REGISTRY-}accountserviceapi
    build:
      context: .
      dockerfile: AccountService.API/Dockerfile
    env_file:
      - .env           # relative path
    environment:
      - ConnectionStrings__DefaultConnection=${ACCOUNT_SERVICE_CONNECTION}
    ports:
    - "8080:8080"
    - "8081:8081"
```
You may also specify file via command:
```bash
docker-compose --env-file .env up
```
4. Add env. variables to your builder configuration:
```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Configuration.AddEnvironmentVariables();
...
```
5. Apply configuration to your AppDbContext:
```csharp
    public static IServiceCollection AddConnetionStrings(this IServiceCollection services, IConfiguration configuration)
    {
        services.AddDbContext<AccountDbContext>(options =>
            options.UseSqlServer(configuration.GetConnectionString("DefaultConnection")));
            
        return services;
    }
```

Yarashine committed:
Fix pull request and
add private readonly
add interface dbcontext
move business logic from repositories
add pagination like extension
extract common logic to base repository
add transaction
